[Backend] Add API

michellemtchai opened the pull request:
See issue #146 for more details.

**Please merge this PR after PR #151**

michellemtchai committed:
feat(api): create session route
GET /sessions
POST /sessions

michellemtchai committed:
feat(api): create position_templates routes
GET /position_templates
POST /position_templates
GET /available_position_templates

michellemtchai committed:
feat(api): create instructors routes
GET /instructors
POST /instructors

siefkenj commented on api/app/controllers/api/v1/instructors_controller.rb:
```diff
@@ -1,60 +1,32 @@
+        # POST /instructors
+        def create
+            instructor = Instructor.new(instructor_params)
+            if instructor.save
```
What is this?

michellemtchai replied:
Which line are you referring to?

siefkenj replied:
`if instructor.save`

michellemtchai replied:
This checks if the `Instructor` created passes all the validation for the model `Instructor`.

siefkenj replied:
`.save` seems like a really awkward name for `.passes_validation`...

siefkenj replied:
Ah, I understand. `instructor.save` does the action of saving an instructor returning `true` if the save was successful. Can you add a comment to this effect for new rails people?

michellemtchai replied:
Okay. I'll do that.

michellemtchai committed:
fix(api): follow api guideline
change POST /instructors to POST /add_instructor
change POST /position_templates to POST /add_position_template
filter position_templates and positions by session
filter instructors by position

siefkenj commented on api/app/controllers/api/v1/position_templates_controller.rb:
```diff
@@ -0,0 +1,56 @@
+
+        def position_templates_by_session
+            position_templates = []
+            PositionTemplate.order(:id).each do |entry|
```
You can change this to
```
return PositionTemplate.order(:id).select do |entry|
    entry[:session_id] == params[:session_id].to_i
end
```

michellemtchai replied:
Great idea. I'll change it. Thanks.

siefkenj commented on api/app/controllers/api/v1/positions_controller.rb:
```diff
@@ -1,92 +1,83 @@
+        # POST /positions
+        def create
+            req_keys = [:position_code, :position_title]
+            if Session.exists?(id: params[:session_id])
```
I think it makes code more readable to have early returns instead of nested `if`s.

e.g.,
```
if not Session.exists?(id: params[:session_id])
    render json: { status: 'error', message: 'Invalid session_id', payload: {} }
    return
end
if not (req_keys & params.keys.map(&:to_sym) == req_keys)
    render json: { status: 'error',  message: 'Missing position_code or position_title', payload: {} }
    return
end

position = Position.new(position_params)
...
```

michellemtchai replied:
I've removed the nested `if`s.

siefkenj commented on api/app/controllers/api/v1/instructors_controller.rb:
```diff
@@ -1,61 +1,52 @@
+            if Position.exists?(id: params[:position_id])
+                render json: { status: 'success', message: '', payload: instructors_by_position }
+            else
+                render json: { status: 'error', message: 'Invalid position_id', payload: {} }
```
What if `:position_id` isn't provided because this is used from a different route?

michellemtchai replied:
I've made a new route `GET /api/v1/instructors` for the `index` function. Is this what you mean?

siefkenj replied:
Yep, that's what I mean.

michellemtchai committed:
fix(api): change instructor.create to instructor.new
made this change because .create also include .save, so it was redundant
update erd to match uniqueness

michellemtchai committed:
fix(api): simplify filter list
use .select to filter out db entries instead of .each

michellemtchai committed:
fix(api): create GET /instructors
create alternative to getting instructor than GET /positions/:position_id/instructors

michellemtchai committed:
fix(api): remove nested ifs :create
removed the nested if in the following controllers:
position_template
position
session
instructor

siefkenj commented on api/app/controllers/api/v1/positions_controller.rb:
```diff
@@ -1,92 +1,79 @@
+                render json: { status: 'error', message: 'Invalid session_id', payload: {} }
+                return
+            end
+            if (req_keys & params.keys.map(&:to_sym)) != req_keys
```
https://api.rubyonrails.org/classes/ActionController/Parameters.html#method-i-require looks to be the standard method to require particular `params`. Is there a reason you didn't use it?

michellemtchai replied:
Actually, I did not know about this function. I'll change it to `require`. Thanks.

siefkenj commented on api/app/controllers/api/v1/positions_controller.rb:
```diff
@@ -1,92 +1,79 @@
+                return
+            end
+            position = Position.new(position_params)
+            if position.save
```
I like this the other way around, too.
```
if not position.save
   ...
end
...rest of code
```

michellemtchai replied:
I've move the `not position.save` to the top. I've also added a validation for `position_data_for_matching` and `position_data_for_ad`. Should I move that up as well?

siefkenj commented on api/app/controllers/api/v1/sessions_controller.rb:
```diff
@@ -0,0 +1,38 @@
+
+        # POST /sessions
+        def create
+            if not params.include?(:name)
```
See previous comment about required params

michellemtchai replied:
Done.

michellemtchai committed:
fix(db): clean up if blocks & use params.require
moved the if block to the top of the functions for :create
use params.require for required fields
added validation for position, position_data_for_matching, and position_data_for_ad

siefkenj commented on api/app/controllers/api/v1/positions_controller.rb:
```diff
@@ -1,92 +1,92 @@
+            end
+            position = Position.new(position_params)
+            if not position.save # does not pass Position model validation
+                render json: { status: 'error', message: position.errors, payload: {} }
```
Need a `return` here

michellemtchai replied:
Great catch. Thanks. I'll fix that.

siefkenj commented on api/app/controllers/api/v1/positions_controller.rb:
```diff
@@ -1,92 +1,92 @@
+                return nil
+            elsif ad.save and not matching.save
+                return matching.errors
+            elsif not ad.save and matching.save
```
Does this code re-execute `*.save` a bunch of times? Will it save items to the database multiple times??

michellemtchai replied:
I don't know if this re-execute a bunch of times, so I'll just save the outcome to a local variable.

siefkenj commented on api/app/controllers/api/v1/position_templates_controller.rb:
```diff
@@ -0,0 +1,55 @@
+                    message: 'Invalid session_id', payload: position_templates_by_session }
+                return
+            end
+            if not params.require(:offer_template)
```
`params.require` documentation says it raises an error if the requirement is not met, so I don't think this is the right way to use it.

michellemtchai replied:
Okay. I've gotten rid of this and wrote a rescue from `ParameterMissing` instead.

michellemtchai committed:
feat(api): setup all routes from plan
GET /applicants
POST /applicants
GET /applications
GET /sessions/:session_id/applications
POST /sessions/:session_id/applications
GET /assignments
GET /positions/:position_id/assignments
POST /positions/:position_id/assignments
POST /applications/:application_id/add_preference
GET /positions/:position_id/assignments
POST /positions/:position_id/assignments
GET /assignments/:assignment_id/active_offer
POST /assignments/:assignment_id/add_wage_chunk
POST /wage_chunks/:wage_chunk_id/add_reporting_tag
POST /creat_offer
POST /email_offer
POST /ta/offers/:offer_id/respond_to_offer

michellemtchai committed:
fix(api): use local variable instead calling .save multiple time
use ad_save and matching_save to get outcome of save

michellemtchai committed:
fix(api): use rescue from ParameterMissing
render json with status error whenever params.require fails

siefkenj commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -0,0 +1,77 @@
+            params[:application_id] = application[:id]
+            message = valid_matching
+            if not message
+                render json: { status: 'error', message: '', payload: application }
```
should this be `status: 'success'`?

michellemtchai replied:
Yes. It should be. Thanks for catching that.

siefkenj commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -0,0 +1,77 @@
+                render json: { status: 'error', message: application.errors, payload: {} }
+            end
+            params[:application_id] = application[:id]
+            message = valid_matching
```
This is a confusing name...can you come up with something better?

michellemtchai replied:
Is `valid_applicant_matching_data` a good alternative? Or should I come up with something else?

michellemtchai committed:
fix(api): fix status render mistake & make better function name
fix the status for application :create so that it has a success status
change valid_matching into valid_applicant_matching_data

michellemtchai committed:
feat(api): added more routes for applicants & wage_chunks
GET /applicants
GET /sessions/:session_id/applicants
GET /wage_chunks
GET /assignments/:assignment_id/wage_chunks

michellemtchai commented:
I've added the routes:
```
GET /sessions/:session_id/applicants
GET /assignments/:assignment_id/wage_chunks
```

michellemtchai committed:
feat(api): make helper functions in response.rb
created render_success(payload)
created render_error(message, payload={})
created invalid_primary_key(table, params_key, payload={})

michellemtchai committed:
fix(api): make invalid_primary_key to invalid_id
make function name shorter for easier inline if returns

michellemtchai committed:
fix(api): remove typo in valid_ad_and_matching
change matching_save = matching_save to matching_save = matching.save

michellemtchai committed:
fix(api): create wage_chunks when create assignments has hours
add condition to create wage_chunk given hours

michellemtchai committed:
fix(api): change WageChunk.new to ReportingTag.new in reporting_tags_controller
fix typo

michellemtchai committed:
feat(api): implement update for controllers
PUT/PATCH routes were made for the following controllers:
applicants
applications
instructors
position_preferences
position_templates
positions
reporting_tags
sessions
wage_chunks

michellemtchai committed:
fix(api): destroy db entry when entry POST invalid
destroy db entry when entry fails validation

siefkenj commented:
Is this ready to merge?

michellemtchai commented:
I guess it is ready to merge if you don't see any glaringly missing route/error.

siefkenj commented:
I'll go ahead and merge. It looks mostly good. I am sure some bugs will be revealed when we have seed data and unit tests, but since we don't have a full product yet, let's push forward for rapid progress!
